src/StatsdClient/MetricSerializer.cs

```
@@ -0,0 +1,245 @@
+            constantTags = constantTags ?? EmptyStringArray;
+            tags = tags ?? EmptyStringArray;
+
+            if (constantTags.Length == 0 && tags.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            var allTags = constantTags.Concat(tags);
+            string concatenatedTags = string.Join(",", allTags);
+            return $"|#{concatenatedTags}";
+        }
+
+        private static string TruncateOverage(string str, int overage)
+        {
+            return str.Substring(0, str.Length - overage);
+        }
+
+        public abstract class Metric
+        {
+            private static readonly Dictionary<MetricType, string> _commandToUnit = new Dictionary<MetricType, string>
+                                                                {
+                                                                    { MetricType.Counting, "c" },
+                                                                    { MetricType.Timing, "ms" },
+                                                                    { MetricType.Gauge, "g" },
+                                                                    { MetricType.Histogram, "h" },
+                                                                    { MetricType.Distribution, "d" },
+                                                                    { MetricType.Meter, "m" },
+                                                                    { MetricType.Set, "s" },
+                                                                };
+
+            public static string GetCommand<T>(MetricType metricType, string prefix, string name, T value, double sampleRate, string[] constantTags, string[] tags)
+            {
+                string full_name = prefix + name;
+                string unit = _commandToUnit[metricType];
+                var allTags = ConcatTags(constantTags, tags);
+
+                return string.Format(
```

Are we not concerned with overage for the metric type? Highly unlikely, but that tag array could potentially get large. We can't truncate, but we could raise an exception.

Also, maybe the bufferize worker handler will already complain, but I'm not sure, it didn't seem to me at a glance.

Let me know if anything is unclear from my side.

---

https://github.com/DataDog/dogstatsd-csharp-client/blob/master/src/StatsdClient/Bufferize/StatsBufferize.cs#L67-L70  handles when a metric size is too big.